hail/src/main/scala/is/hail/annotations/Region.scala

```diff
@@ -45,13 +49,53 @@ final class Region() extends NativeBase() {
   @native def nativeAllocate(n: Long): Long
   @native def nativeReference(r2: Region): Unit
   @native def nativeRefreshRegion(): Unit
-  
+
+  @native def nativeGetNumParents(): Int
+  @native def nativeSetNumParents(n: Int): Unit
+  @native def nativeSetParentReference(r2: Region, i: Int): Unit
+  @native def nativeGetParentReferenceInto(r2: Region, i: Int): Region
+  @native def nativeClearParentReference(i: Int): Unit
+
   final def align(a: Long) = nativeAlign(a)
   final def allocate(a: Long, n: Long): Long = nativeAlignAllocate(a, n)
   final def allocate(n: Long): Long = nativeAllocate(n)
 
-  final def reference(other: Region): Unit = nativeReference(other)
+  private var explicitParents: Int = -1
+
+  final def reference(other: Region): Unit = {
+    assert(explicitParents < 0, s"can't use 'reference' if you're explicitly setting Region dependencies")
+    nativeReference(other)
+  }
+
   final def refreshRegion(): Unit = nativeRefreshRegion()
+
+  def setNumParents(n: Int): Unit = {
+//    assert(nativeGetNumParents() == 0, s"Can't explicitly set Region dependencies if references already exist")
```

Comment:
Why is this disabled?

Reply:
I want to be able to expand the number of parents (but not shrink it). I'll fix the assertion, enable it, and add a test.